### PR TITLE
fix(python): Handle Duplicative Import Names

### PR DESCRIPTION
**Pull request by noanflaherty**
## Description
Sibling PR: https://github.com/fern-api/fern/pull/5365

The goal of this PR is to handle auto-aliasing of imports when there are naming conflicts.

## Changes Made
- Track ref name overrides on the `Writer`
- Build up the map of overrides as part of `PythonFile`'s `write` method

## Testing
- [X] Unit tests added/updated

### CODE REVIEW

File: generators/python-v2/ast/src/PythonFile.ts
Comment (noanflaherty) [inline generators/python-v2/ast/src/PythonFile.ts:53]: This method is where the magic happens. Within it, we add state to `Writer`

File: generators/python-v2/ast/src/PythonFile.ts
Comment (noanflaherty) [inline generators/python-v2/ast/src/PythonFile.ts:65]: Wipe the state we had set on the writer just to be safe.

File: generators/python-v2/ast/src/PythonFile.ts
Comment (noanflaherty) [inline generators/python-v2/ast/src/PythonFile.ts:133]: I hope to get rid of the need for this method altogether later. Right now, the data structure we use allows for duplicative refs, but there's no reason it needs to. We can deduplicate upon adding a new ref.

File: generators/python-v2/ast/src/__test__/__snapshots__/PythonFile.test.ts.snap
Comment (noanflaherty) [inline generators/python-v2/ast/src/__test__/__snapshots__/PythonFile.test.ts.snap:117]: I chose a `_n` suffix, but am open to suggestions here.

File: generators/python-v2/ast/src/Reference.ts
Comment (noanflaherty) [inline generators/python-v2/ast/src/Reference.ts:47]: Notably a `Reference` object now gets its name from the `Writer` during its own `write` method.

File: generators/python-v2/ast/src/PythonFile.ts
Comment (noanflaherty) [inline generators/python-v2/ast/src/PythonFile.ts:51]: I hope to get rid of the need for this method altogether later. Right now, the data structure we use for ref tracking allows for duplicative refs, but there's no reason it needs to. We can deduplicate upon adding a new ref.

In any case, this is just a refactor to take existing logic that was there and move it into a helper so its return value can be used in two places directly below.

File: generators/python-v2/ast/src/PythonFile.ts
Comment (noanflaherty) [inline generators/python-v2/ast/src/PythonFile.ts:172]: Here is one of two places where we use the state that was added to the `Writer`.

File: generators/python-v2/ast/src/core/Writer.ts
Comment (noanflaherty) [inline generators/python-v2/ast/src/core/Writer.ts:8]: This shows the state that's now tracked on `Writer`s and helps for interacting with it.

File: generators/python-v2/ast/src/__test__/__snapshots__/PythonFile.test.ts.snap
Comment (noanflaherty) [reply generators/python-v2/ast/src/__test__/__snapshots__/PythonFile.test.ts.snap:117]: @dvargas92495 lmk if you have a preferred naming convention here.

Comment (dvargas92495) [reply generators/python-v2/ast/src/__test__/__snapshots__/PythonFile.test.ts.snap:117]: usually do something module dependent. eg

```python
from .cars import Car as CarsCar
from .vehicles import Car as VehiclesCar
```

File: generators/python-v2/ast/src/PythonFile.ts
Comment (amckinney) [inline generators/python-v2/ast/src/PythonFile.ts:84]: Do we actually need to track the counts? We only care if it's used at all right? Could we just use `reservedNames` instead?

File: generators/python-v2/ast/src/Reference.ts
Comment (amckinney) [inline generators/python-v2/ast/src/Reference.ts:77]: nit: Thoughts `getFullyQualifiedPath`?

File: generators/python-v2/ast/src/core/Writer.ts
Comment (amckinney) [reply generators/python-v2/ast/src/core/Writer.ts:8]: Let's define an `interface` for the value and name it. Something like:

```typescript
interface ImportedName {
  name: string,
  isAlias?: boolean;
}

...

private fullyQualifiedPathsToImportedNames = Record<string, ImportedName> = {};
```

File: generators/python-v2/ast/src/PythonFile.ts
Comment (amckinney) [inline generators/python-v2/ast/src/PythonFile.ts:131]: I'd update the name here - when I think of reserved names I think of keywords like `def`, `__init__`, etc. Maybe just `getUsedNames`?

Comment (noanflaherty) [reply generators/python-v2/ast/src/PythonFile.ts:131]: Yep good call 👍

File: generators/python-v2/ast/src/Reference.ts
Comment (noanflaherty) [reply generators/python-v2/ast/src/Reference.ts:77]: This helper had existed, but yep happy to change as part of this PR.

File: generators/python-v2/ast/src/PythonFile.ts
Comment (noanflaherty) [reply generators/python-v2/ast/src/PythonFile.ts:84]: We do for the current naming convention scheme, but once I adapt to match Vargas's proposal, we won't need to.

File: generators/python-v2/ast/src/core/Writer.ts
Comment (noanflaherty) [reply generators/python-v2/ast/src/core/Writer.ts:8]: Will do

File: generators/python-v2/ast/src/PythonFile.ts
Comment (amckinney) [inline generators/python-v2/ast/src/PythonFile.ts:78]: nit: Should we define an interface for this map value?

File: generators/python-v2/ast/src/Reference.ts
Comment (amckinney) [inline generators/python-v2/ast/src/Reference.ts:77]: nit: I would use `getFullyQualifiedModulePath` and `getFullyQualifiedPath` (since omitting the `Module` component implies its for the reference), but feel free to ignore if you disagree. Not a big deal either way, and it's easy to change later (if ever).

File: generators/python-v2/ast/src/__test__/__snapshots__/PythonFile.test.ts.snap
Comment (amckinney) [inline generators/python-v2/ast/src/__test__/__snapshots__/PythonFile.test.ts.snap:124]: 👏

File: generators/python-v2/ast/src/core/utils.ts
Comment (amckinney) [inline generators/python-v2/ast/src/core/utils.ts:48]: There's a lot of custom regex in here - could we have used some combination of [lodash](https://github.com/lodash/lodash) functions for the same effect? This might be as simple as just calling `upperFirst(camelCase(name))` to generate PascalCase like [this](https://github.com/fern-api/fern/blob/f5529b192d991f3c3a6ec2f0458e409ad5a968b3/packages/cli/generation/ir-generator/src/casings/CasingsGenerator.ts#L38C34-L38C60) (plus potentially some other minor cleanups).

Would you mind looking into this a bit to see if we can remove this altogether?

File: generators/python-v2/ast/src/PythonFile.ts
Comment (amckinney) [inline generators/python-v2/ast/src/PythonFile.ts:101]: Note that if we don't use a simple `_` strategy, the generated names can potentially get _long_ unless you maintain a collision map and assign type names more carefully.

For example, imagine the following module name (which we actually sometimes see in practice):

```python
from .resources.v1.api.vehicles.automobiles import Car as ResourcesV1ApiVehiclesAutomobilesCar
class Car:
    vehicle = ResourcesV1ApiVehiclesAutomobilesCar()
```

For what it's worth, we do this in the original Go generator but only prepend names as needed (until we find a unique name). Given that we traverse the statements in a deterministic order, the import alias assignment is always guaranteed to be unique. You can check out the method docs [here](https://github.com/fern-api/fern/blob/f5529b192d991f3c3a6ec2f0458e409ad5a968b3/generators/go/internal/gospec/scope.go#L60), and [this](https://github.com/fern-api/fern/blob/f5529b192d991f3c3a6ec2f0458e409ad5a968b3/generators/go/internal/gospec/scope_test.go#L17) test case.

**Review by amckinney — APPROVED** [review]
Stamping to unblock, but there's a few things I think we can improve here. I trust you to consider them before merging, but I'll leave you to making a judgement call for your use case.

File: generators/python-v2/ast/src/PythonFile.ts
Comment (noanflaherty) [reply generators/python-v2/ast/src/PythonFile.ts:101]: Perhaps I'm misunderstanding, but I assert the current implementation _does_ only prepend names as needed until it comes up with one that's unique. It goes from right to left, adding one module to the name at a time, until it finds that it's come up with a unique name.

Comment (noanflaherty) [reply generators/python-v2/ast/src/PythonFile.ts:101]: Updating the test to make this clear makes sense. Will do!

**Pull request merged**
